### PR TITLE
Support images in 'review' and `pr` diff viewers

### LINKED ISSUE #1356: Adding gif in PR shows "An error occurred while loading the image"
- Extension version: 2019.9.38342
- VSCode Version: 1.39.0-insider
- OS: macOS 10.14

Steps to Reproduce:

1. Have a PR that adds a gif (like https://github.com/microsoft/vscode-extension-samples/pull/108)
2. Open the Files -> demo.gif from "Changes in pull request"

image

### PR DESCRIPTION
**Pull request by JacksonKearl**
Fixes #1356.

Uses a trick from the builtin git extension to represent images as data URL's, as the `provideTextDocument` API does not handle images. There is a `fileSystemProvider` API that we could use instead, but it ends up being more complicated.

### CODE REVIEW

File: src/common/uri.ts [hunk 1]
```diff
@@ -43,9 +45,34 @@ export interface GitUriOptions {
...
+		const { size, object } = await repository.getObjectDetails(commit, uri.fsPath);
+		const { mimetype } = await repository.detectObjectType(object);
+
+		if (mimetype === 'text/plain') {
+			return;
+		}
+
+		if (ImageMimetypes.indexOf(mimetype) > -1) {
+			const contents = await repository.buffer(commit, uri.fsPath);
+			return Uri.parse(`data:${mimetype};label:${path.basename(uri.fsPath)};description:${commit};size:${size};base64,${contents.toString('base64')}`);
```
Comment (RMacfarlane) [inline src/common/uri.ts]: nit: tslint warning about `path` variable being shadowed, perhaps rename the import

File: src/common/uri.ts [hunk 2]
```diff
@@ -43,9 +45,34 @@ export interface GitUriOptions {
...
+	'image/gif',
+	'image/jpeg',
+	'image/webp',
+	'image/tiff',
+	'image/bmp'
+];
+
+export async function asImageDataURI(uri: Uri, repository: Repository): Promise<Uri | undefined> {
+	try {
+		const { commit } = JSON.parse(uri.query);
```
Comment (RMacfarlane) [inline src/common/uri.ts]: for `pr` scheme uris, the query has `headCommit`, `baseCommit`, `isBase`. We could try to use those or just short-circuit for non-review scheme uris

File: src/commands.ts [hunk 3]
```diff
@@ -160,7 +162,18 @@ export function registerCommands(context: vscode.ExtensionContext, prManager: Pu
 		if (isPartial) {
 			vscode.window.showInformationMessage('Your local repository is not up to date so only partial content is being displayed');
 		}
-		vscode.commands.executeCommand('vscode.diff', parentFilePath, filePath, fileName, opts);
+
+		// if this is an image, encode it as a base64 data URI
+		const imageDataURI = await asImageDataURI(parentFilePath, prManager.repository);
+		const originalURI = imageDataURI || parentFilePath;
+
+		if (fileChangeNode.status === GitChangeType.ADD) {
```
Comment (RMacfarlane) [inline src/commands.ts]: I'm not sure if we should do this or not - on the one hand it adds a bunch of screen real estate when viewing these files, but it also makes it less obvious at first glance if they are an add or delete. My preference is for the old way

File: src/common/uri.ts (hunk 1 above)
Comment (JacksonKearl) [reply src/common/uri.ts]: Done!

File: src/commands.ts (hunk 3 above)
Comment (JacksonKearl) [reply src/commands.ts]: Moved to using a transparent pixel URI, which is what the builtin diff view does.

File: src/common/uri.ts (hunk 2 above)
Comment (JacksonKearl) [reply src/common/uri.ts]: Changed this and the call up a bit, now we can view added, modified, and deleted images in both `pr` and `review` schemes.

**Pull request merged**